guardian/support-service-lambdas
# feat: implement weekly health check flow for the Salesforce Disaster Recovery state machine

## What does this change?

[Trello](https://trello.com/c/stuJuxvX/771-implement-health-check-for-re-syncing-procedure-system)

- Create new application `salesforce-disaster-recovery-health-check` in charge of running weekly health checks on the Salesforce Disaster Recovery state machine:

  - An AWS rule is triggered every Monday at 6AM

  - A function starts an execution of the state machine:
    - the query filters only active accounts created yesterday (e.g. `SELECT Id, Zuora__Zuora_Id__c, Zuora__Account__c, Contact__c FROM Zuora__CustomerAccount__c WHERE CreatedDate = YESTERDAY AND Zuora__Status__c = 'Active'`)

    - A custom execution name is passed so that the state machine can detect whether the execution is a health check (e.g. `health-check-20240411T111548720Z`)
  - the same function sends a notification if the state machine is not healthy (same SNS topic as main application)

- Amend the main `salesforce-disaster-recovery` state machine:

  - Improve state names:
    - `ConstructEmailData` -> `ConstructNotificationData`

    - `SendProcedureSummaryEmail` -> `SendCompletionNotification`

  - Add choice state to skip sending the completion notification if the execution is for health check. The state machine uses intrinsic functions to extract the execution name from the context (e.g. `Condition.stringMatches('$$.Execution.Name', 'health-check-*')`)

## Why are you doing this?

This new testing application makes sure the PROD Salesforce Disaster Recovery re-syncing procedure is available at all times (e.g. expired auth credentials or amended Zuora permission).

Thanks to @johnduffell  for the idea.

## Architecture

<div align="center">
  <table>
    <tr>
      <td align="center" valign="middle">
        <img width="697" alt="Screenshot 2024-04-11 at 13 04 55" src="https://github.com/guardian/support-service-lambdas/assets/39066365/7c3a675c-c0d2-4381-93fd-abb614070bcc">
      </td>
      <td align="center" valign="middle">
       <img width="573" alt="Screenshot 2024-04-11 at 13 05 20" src="https://github.com/guardian/support-service-lambdas/assets/39066365/705dad6f-8623-4017-b820-f2b25bf75b49">
      </td>
    </tr>
  </table>
</div>

## How to test

Execute the CODE lambda function from the AWS console [here](https://eu-west-1.console.aws.amazon.com/lambda/home?region=eu-west-1#/functions/salesforce-disaster-recovery-health-check-CODE?tab=testing).

Find and monitor the state machine execution [here](https://eu-west-1.console.aws.amazon.com/states/home?region=eu-west-1#/statemachines/view/arn%3Aaws%3Astates%3Aeu-west-1%3A865473395570%3AstateMachine%3Asalesforce-disaster-recovery-CODE).

<img width="789" alt="Screenshot 2024-04-11 at 15 24 28" src="https://github.com/guardian/support-service-lambdas/assets/39066365/303751f9-2826-4f9d-aee3-28cbb0662755">

<div align="center">
<img width="402" alt="Screenshot 2024-04-11 at 15 26 22" src="https://github.com/guardian/support-service-lambdas/assets/39066365/acc7cfb2-6366-4662-8e3c-f31f6a3c6b2c">
</div>
